autogpt/memory/sqlitemem.py

```diff
@@ -0,0 +1,368 @@
+            old_embeddings.rename(old_embeddings.with_suffix(".zarr.old"))
+        else:
+            shutil.rmtree(old_embeddings, ignore_errors=True)
+
+    def add(self, data: str) -> str:
+        """
+        Adds a data point to the memory.
+
+        Args:
+            data: The data to add.
+
+        Returns: Message indicating that the data has been added.
+        """
+        # Create the data embedding
+        embedding = get_ada_embedding(data)
+        embedding = np.expand_dims(embedding, axis=0)
+
+        # Get the embedding index (index of the next vector is the same as the number of vectors)
+        embedding_index = self.embeddings.shape[0]
+
+        # Add data to the database
+        with self.sql_conn as conn:
+            try:
+                conn.execute(
+                    f"INSERT INTO {self.index} (id, data, timestamp) VALUES(?, ?, CURRENT_TIMESTAMP);",
+                    (embedding_index, data),
+                )
+            except sqlite3.IntegrityError as err:
+                if "UNIQUE constraint failed:" in str(err):
+                    return "Trying to add duplicate data to memory, ignoring."
+                raise err
+
+        # Store the embedding
+        self.embeddings.append(embedding)
+        return f"Inserting data into memory at index: {embedding_index}:\ndata: {data}"
+
+    def get(self, data: str) -> Optional[List[Any]]:
+        """
+        Gets the data from the memory that is most relevant to the given data.
+
+        Args:
+            data: The data to compare to.
+
+        Returns: The most relevant data.
+        """
+        return self.get_relevant(data, 1)
+
+    def clear(self) -> str:
+        """
+        Clears the memory.
+
+        Returns: A message indicating that the memory has been cleared.
+        """
+        # Wipe the embeddings for index
+        self.embeddings = None
+        shutil.rmtree(self.embeddings_path, ignore_errors=True)
+        self._create_embeddings()
+        self.embeddings = zarr.open(self.embeddings_path, mode="a")
+
+        # Wipe the table for index
+        with self.sql_conn as conn:
+            conn.execute(f"DROP TABLE IF EXISTS {self.index};")
+        self._create_table()
+
+        return "Obliviated"
+
+    def get_relevant(self, data: str, num_relevant: int = 5) -> Optional[List[Any]]:
+        """
+        Returns all the data in the memory that is relevant to the given data.
+        Args:
+            data: The data to compare to.
+            num_relevant: The number of relevant data to return.
+
+        Returns: A list of the most relevant data.
+        """
+        # If there are no data points in the memory, return empty list
+        if self.embeddings.shape[0] == 0:
+            return []
+
+        # Create the data embedding
+        embedding = get_ada_embedding(data)
+        embedding = np.expand_dims(embedding, axis=0)
+
+        # Get the nearest neighbors
+        nbrs = NearestNeighbors(
```

Comment:
Why bring in sklearn for nearest neighbors here?
We already have some metrics for doing look up of relevant embeddings, see local cache.
Or a simple SVM performs as well or better than the dot product performs well and is computationally efficient. This is presumably a brute force algorithm that won't scale well.